src/babylon/tools/data/CuckooHash.tsx

```diff
@@ -0,0 +1,369 @@
+            this._type.setBufferValue(this._tableView, pos, 0, 0);
+            --this._setElements;
+            if (this._set(lo, hi, 0, []) === false) {
+              ++retries;
+              tmp = this.getLastInsertError();
+              break;
+            }
+          }
+        }
+      }
+
+      if (tmp.lo === 0 && tmp.hi === 0) {
+        return true;
+      }
+    }
+    return this._resize();
+  }
+
+  private _resize(): boolean {
+    const level = this._sizes.level;
+
+    if (level + 1 < Cuckoo._sizeLookup.length) {
+      this._sizes = Cuckoo._sizeLookup[level + 1];
+
+      // Create new empty ArrayBuffer (can't resize existing one)
+      let newTable = new ArrayBuffer(this._type.bytePerElement * this._sizes.maxsize);
+      let newTableView = new DataView(newTable);
+
+      // Copy old buffer dumb way until we have ArrayBuffer.transfer()
+      for (let i = 0; i < this._tableView.byteLength; ++i) {
+        newTableView.setUint8(i, this._tableView.getUint8(i));
+      }
+      this._table = newTable;
+      this._tableView = newTableView;
+
+      this._updateTexture();
+
+      // Rehash all elements
+      return this._rehash();
+    }
+
+    // Can't enlarge hash table - already at max size
+    return false;
+  }
+
+  private _updateTexture(): void {
+    const engine = this._scene.getEngine();
+    const gl = engine._gl as WebGL2RenderingContext;
+
+    if (this._texture._texture !== undefined) {
+      this._texture.releaseInternalTexture();
+    }
+
+    this._texture._texture = engine.createDynamicTexture(this._sizes.texwidth, this._sizes.texheight, false,
+      BABYLON.Texture.NEAREST_SAMPLINGMODE);
+
+    gl.bindTexture(gl.TEXTURE_2D, this._texture._texture);
+
+    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
+    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
+    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
+    gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+    this._type.texImage2D(gl, this._sizes.texwidth, this._sizes.texheight, undefined);
+
+    gl.bindTexture(gl.TEXTURE_2D, null);
+
+    this._texture._texture.isReady = true;
+  }
+
+  private _set(lo: number, hi: number, mode: number, history: { pos: number, lo: number, hi: number }[]): boolean {
+    const pos = this._hashfunctions[mode].compute(lo, hi) % this._sizes.maxsize;
+    const [loOld, hiOld] = this._getValue(pos);
+
+    if (loOld === 0 && hiOld === 0) {
+      this.clearLastInsertError();
+
+      this._setValue(pos, lo, hi);
+      ++this._setElements;
+      if (this.loadFactor > 0.5 && this._sizes.level + 1 < Cuckoo._sizeLookup.length) {
+        return this._resize();
+      }
+      return true;
+    } else if (loOld === lo && hiOld === hi) {
+      return true;
+    } else {
+      // Check for circuit
+      if (history.find((el) => { return el.pos === pos && el.hi === hi && el.lo === lo; })) {
```

Comment:
We should probably see how often this is occurring. Also, collect statistics on how large history actually gets.
If history is actually very very large, we should consider using a Dictionary instead.